cni/pkg/nodeagent/podcgroupns.go

```diff
@@ -0,0 +1,370 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodeagent
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/fs"
+	"path"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/istio/pkg/util/sets"
+)
+
+type PodToNetns map[types.UID]NetnsCloser
+
+func (p PodToNetns) Close() {
+	for _, netns := range p {
+		netns.Close()
+	}
+}
+
+type PodNetnsFinder interface {
+	FindNetnsForPods(filter sets.Set[types.UID]) (PodToNetns, error)
+}
+
+type PodNetnsProcFinder struct {
+	proc fs.FS
+}
+
+func NewPodNetnsProcFinder(proc fs.FS) *PodNetnsProcFinder {
+	return &PodNetnsProcFinder{proc: proc}
+}
+
+func isNotNumber(r rune) bool {
+	return r < '0' || r > '9'
+}
+
+func (p *PodNetnsProcFinder) FindNetnsForPods(filter sets.Set[types.UID]) (PodToNetns, error) {
+	/*
+		for each process, find its netns inode,
+		if we already seen the inode, skip it
+		if we haven't seen the inode, check the process cgroup and see if we
```

Comment:
The current lookup is  pod-ip -> interface.
Also we get the netns id, not fd; that means that we need to map the id back to the fd by enumerating all netns /var/run/netns (see `getLinkWithDestinationOf`).

This code is pod-uid->netns fd

I think lookup by uid has the advantage of less chance of error / races with ips changing.

If you are concerned with this method's reliability, maybe we can do it first, and fallback to `getLinkWithDestinationOf`?

Reply:
Are we keeping the podip->interface code or did we remove it?

_If we already have that_ and rely on it, I think we can get the netns more efficiently (https://github.com/istio/istio/pull/48284/files#diff-ef3ae91920da25719784a8d0f788bcceaaa11eba8408ff8728914fa33f9fafe4R90, which is interface->netns). If we are ditching that code entirely anyways that may be less applicable

Reply:
Using the interface in the lookup has the disadvantage (as noted in that code in a comment) of not working in all environments and being reliant on specific assumptions about how host node networking is configured.


This approach does not have that limitation and is going to work on any Linux node with an OCI-compliant container runtime. It's similar to what Cilium and others do, and they do it for that reason.

Reply:
The exceptions for crio, etc in the cgroups regex stuff makes me a bit skeptical its actually going to work on any OCI-compliant runtime? To be fair, the fact others are using similar mechanisms gives us some "herd immunity" here, but still seems both are heuristic based methods

Reply:
> The exceptions for crio, etc in the cgroups regex stuff makes me a bit skeptical its actually going to work on any OCI-compliant runtime?


The overall mechanism will work for all of them, more broadly than the interface-based approach will.

Regex does need to be different for some, because while pod ID is by-definition linked to container ID by way of cgroup _somehow_ (either filename or path), the representation for that is not standardized, unfortunately.

Reply:
I don't think there's a way to do this none-heuristically, Unless you get access to the CRI itself. 
Containerd and CRI-O will provide namespace information in verbose mode. Though this is **not** [guaranteed](https://github.com/kubernetes/cri-api/blob/124dcc10af511d643d703abfea24a217795783ca/pkg/apis/runtime/v1/api.proto#L568) by the spec (as it is meant for debugging).

The CRI has no permissions model, and this gives a lot of "power" to the node agent, which may not be ideal.
In addition it requires knowing the location and getting access to the CRI socket.

So I feel like all our options are heuristics. We can do more than one to get more coverage. IMO the current approach is a reasonable place to start